src/anomalib/metrics/spro.py

```diff
@@ -0,0 +1,166 @@
+
+    def __init__(self, threshold: float = 0.5, saturation_config: dict | None = None, **kwargs) -> None:
+        super().__init__(**kwargs)
+        self.threshold = threshold
+        self.saturation_config = saturation_config
+        self.add_state("score", default=torch.tensor(0.0), dist_reduce_fx="sum")
+        self.add_state("total", default=torch.tensor(0), dist_reduce_fx="sum")
+
+    def update(self, predictions: torch.Tensor, masks: list[torch.Tensor]) -> None:
+        """Compute the SPRO score for the current batch.
+
+        Args:
+            predictions (torch.Tensor): Predicted anomaly masks.
+            masks (list[torch.Tensor]): Ground truth anomaly masks with original height and width. Each element in the
+                list is a tensor list of masks for the corresponding image.
+
+        Example:
+            To update the metric state for the current batch, use the ``update`` method:
+
+            >>> spro.update(preds, labels)
+        """
+        score, total = spro_score(
+            predictions,
+            masks,
+            threshold=self.threshold,
+            saturation_config=self.saturation_config,
+        )
+        self.score += score
+        self.total += total
+
+    def compute(self) -> torch.Tensor:
+        """Compute the macro average of the SPRO score across all masks in all batches.
+
+        Example:
+            To compute the metric based on the state accumulated from multiple batches, use the ``compute`` method:
+
+            >>> spro.compute()
+            tensor(0.5433)
+        """
+        if self.total == 0:  # only background/normal images
+            return torch.Tensor([1.0])
+        return self.score / self.total
+
+
+def spro_score(
+    predictions: torch.Tensor,
+    targets: list[torch.Tensor],
+    threshold: float = 0.5,
+    saturation_config: dict | None = None,
+) -> torch.Tensor:
+    """Calculate the SPRO score for a batch of predictions.
+
+    Args:
+        predictions (torch.Tensor): Predicted anomaly masks.
+        targets: (list[torch.Tensor]): Ground truth anomaly masks with original height and width. Each element in the
+            list is a tensor list of masks for the corresponding image.
+        threshold (float): When predictions are passed as float, the threshold is used to binarize the predictions.
+        saturation_config (dict): Saturations configuration for each label (pixel value) as the keys.
+            Defaults: ``None`` (which the score is equivalent to PRO metric, but with the 'region' are
+            separated by mask files.
+
+    Returns:
+        torch.Tensor: Scalar value representing the average SPRO score for the input batch.
+    """
+    # Add batch dim if not exist
+    if len(predictions.shape) == 2:
+        predictions = predictions.unsqueeze(0)
+
+    # Resize the prediction to have the same size as the target mask
+    predictions = torch.nn.functional.interpolate(predictions.unsqueeze(1), targets[0].shape[-2:])
```

Comment:
@jpcbertoldo Ah yes, I agree the resizing process should be outside of the metric. Regarding this resizing process, there is a thing that I'm not really sure about the standard or correct procedure to handle this.

(a) Should we resize the predictions to match the size of the original mask? (b) Or should we resize the original mask to match the size of the prediction?

For the existing `mask`, it seems the mask is the one being resized to match the prediction, i.e., it depends on the `image_size` parameter in the config, right? The resizing process happens in `__getitem__`.
https://github.com/openvinotoolkit/anomalib/blob/2cdeb82eb79348a23325fb3d873862775366faa2/src/anomalib/data/base/dataset.py#L133

However, I noticed the way MVTec LOCO is evaluated [here](https://github.com/nelson1425/EfficientAD) (I'm not sure if this is the common procedure or not): the prediction is resized back to the original size during inference, i.e., following the size of the original mask. Then, resized predictions are evaluated on the original mask size.

If (b), maybe we can simply remove this interpolation and let the transform in `__get_item__` transform the `masks` to match the prediction size. If (a), maybe we need to resize the prediction back to the original size in this `post_process`, along with providing a warning to let the user know the image size will be resized back to the original mask during evaluation.
https://github.com/openvinotoolkit/anomalib/blob/main/src/anomalib/callbacks/post_processor.py#L97

@samet-akcay @ashwinvaidya17 @blaz-r

Reply:
>  I would propose to use the same convention for the SPRO metric as for the other metrics, which is to resize the masks and evaluate at the user-defined image size. We can make the switch to evaluation at the original size in a separate PR.

@djdameln I agree with this. I will resize the `masks` and evaluate at the user-defined image size 👍

Reply:
As @djdameln said, it's probably a good idea to make it simple for the sake of finishing this PR, but that's not ideal. In our latest paper (AUPIMO http://arxiv.org/abs/2401.01984) we specifically talk about this actually : ) because authors often use different ways to resize the masks and that creates artifacts.

Reply:
Let's add a todo note + GitHub issue somewhere so that we don't forget it.